File: 4392-shared/project_1/src/python/solver.py
from enum import Enum

class Move(Enum):
	right = ("right", lambda pos: (pos[0] + 1, pos[1]))
	left = ("left", lambda pos: (pos[0] - 1, pos[1]))
	up_right = ("up-right", lambda pos: (pos[0], pos[1] - 1))
	up_left = ("up-left", lambda pos: (pos[0] - 1, pos[1] - 1))
	down_right = ("down-right", lambda pos: (pos[0] + 1, pos[1] + 1))
	down_left = ("down-left", lambda pos: (pos[0], pos[1] + 1))

def get_col_row(idx):
	row = 0
	while idx > row:
		row += 1
		idx -= row
	return idx, row

def get_idx(col_row):
	idx = col_row[0]
	row = col_row[1]

	while row > 0:
		idx += row
		row -= 1

	return idx

class Peg:
	def __init__(self, col_row, filled):
		self.col = col_row[0]
		self.row = col_row[1]
		self.filled = filled

	def copy(peg):
		return Peg((peg.col, peg.row), peg.filled)

class BoardState:
	def __init__(self, rows, start, pegs = None, moves = []):
		self.rows = rows
		self.start = start
		self.moves = moves
		self.hash = None

		#print("Init BoardState with rows=" + str(self.rows) + 
		#	", start=" + str(get_col_row(self.start)) + ", idx=" + str(get_idx(get_col_row(self.start))) + ", moves=" + str(self.moves))

		if pegs is None:
			self.pegs = [Peg(get_col_row(i), i != start) for i, _ in enumerate([1 for i in range(1, rows + 1) for j in range(1, i + 1)])]
		else:
			self.pegs = [Peg.copy(peg) for peg in pegs]

	def get_all_starting_states(rows):
		boards = []
		holes = int(rows * (rows + 1) / 2)

		for start in range(holes):
			boards.append(BoardState(rows, start))

		return boards

	def __eq__(self, that):
		for i, peg in enumerate(self.pegs):
			if peg.filled != that.pegs[i].filled:
				return False
		return True

	def __hash__(self):
		if self.hash:
			return self.hash

		value = 0
		for i, peg in enumerate(self.pegs):
			value += (1 << i) * (1 if peg.filled else 0)

		self.hash = value
		return value

	def make_move(self, peg_idx, move):
		def is_oob(pos):
			return pos[0] < 0 or pos[1] < 0 or pos[0] >= self.rows or pos[1] >= self.rows or pos[0] > pos[1]

		pos = get_col_row(peg_idx)

		if is_oob(pos) or not self.pegs[peg_idx].filled:
			return None

		next = move.value[1](pos)
		if is_oob(next) or not self.pegs[get_idx(next)].filled:
			return None

		jump = move.value[1](next)
		if is_oob(jump) or self.pegs[get_idx(jump)].filled:
			return None

		board = BoardState(self.rows, self.start, self.pegs, self.moves + [(pos, move.value[0])])
		board.pegs[peg_idx].filled = False
		board.pegs[get_idx(next)].filled = False
		board.pegs[get_idx(jump)].filled = True

		return board

	def __repr__(self):
		string = repr(list(map(lambda peg: 1 if peg.filled else 0, self.pegs)))
		string += "\nstart: " + str(get_col_row(self.start))
		string += "\nhistory: " + str(self.moves)
		return string


def solve(n):
	print("Solving with %d" % n)

	states = BoardState.get_all_starting_states(n)
	explored = set()

	while len(states) > 0:
		state = states.pop(0)

		if state in explored:
			continue

		explored.add(state)

		found_move = False
		for i, _ in enumerate(state.pegs):
			for move in Move:
				next_state = state.make_move(i, move)
				if next_state:
					states.append(next_state)
					found_move = True

		if not found_move:
			return state

	return None

if __name__ == "__main__":
	import sys
	solution = solve(int(sys.argv[1]))
	start_pos = get_col_row(solution.start)
	print(str((start_pos[1], start_pos[0])) + " Start")
	for move in solution.moves:
		print(str((move[0][1], move[0][0])) + " " + str(move[1]))

File: 4392-shared/project_2/src/ocaml/solver.ml
open Printf
open Async.Std
open Core

type move = Start | Right | Left | Up_Right | Up_Left | Down_Right | Down_Left;;

type board_move = {
	move: move;
	idx: int;
};;

type board_state = {
	pegs: int array;
	history: board_move list;
};;

let to_col_row idx =
	let rec to_col idx row =
		if idx > row then
			to_col (idx - row - 1) (row + 1)
		else
			(idx, row)
	in
	to_col idx 0;;

let rec to_idx (col, row) =
	if row > 0 then
		to_idx (col + row, row - 1)
	else
		col;;

let get_move_pos (col, row) move =
	match move with
		| Start -> raise (Failure "This should never happen!")
		| Right -> (col + 1, row)
		| Left -> (col - 1, row)
		| Up_Right -> (col, row - 1)
		| Up_Left -> (col - 1, row - 1)
		| Down_Right -> (col + 1, row + 1)
		| Down_Left -> (col, row + 1);;

let make_move state idx move acc =
	if state.pegs.(idx) == 0 then
		acc
	else
		let is_oob (col, row) = col < 0 || row < 0 || (to_idx (col, row)) >= Array.length state.pegs || col > row in
		let pos = to_col_row idx in
		let next_pos = get_move_pos pos move in
		let next_idx = to_idx next_pos in
		let next_next_pos = get_move_pos next_pos move in
		let next_next_idx = to_idx next_next_pos in

		if is_oob pos || is_oob next_pos || is_oob next_next_pos ||
			state.pegs.(next_idx) == 0 || state.pegs.(next_next_idx) == 1 then
			acc
		else
			let new_pegs = Array.copy state.pegs in
			let () = new_pegs.(idx) <- 0; new_pegs.(next_idx) <- 0; new_pegs.(next_next_idx) <- 1 in
			let new_history = {move = move; idx = idx} :: state.history in
			(Deferred.return ({ pegs = new_pegs; history = new_history })) :: acc;;

let get_next_states state =
	let rec try_peg idx =
		if idx < Array.length state.pegs then
			List.rev_append (make_move state idx Right
								(make_move state idx Left
								(make_move state idx Up_Right
								(make_move state idx Up_Left
								(make_move state idx Down_Right
								(make_move state idx Down_Left []))))))
							(try_peg (idx + 1))
		else
			[]
	in
	try_peg 0;;
let rec to_def_arr llist =
	(match llist with
		| [] -> []
		| h::t -> (Deferred.return h) :: (to_def_arr t)
	)


let solve rows =
	let () = printf "Solving with %d\n" rows in
	let peg_num = rows * (rows + 1) / 2 in
	let states =
		to_def_arr(Array.to_list (Array.init peg_num (fun i -> { pegs = Array.init peg_num (fun j -> if i == j then 0 else 1); history = [{ move = Start; idx = i }]})))
	in
	let rec solve_state states =
		match states with
			| [] -> raise (Failure "This should never happen!")
			| state :: states' -> Deferred.bind state
			(fun s ->
				let next_states = get_next_states s in
				if List.length next_states == 0 then
					Deferred.return s
				else
					solve_state (List.rev_append (List.rev states') next_states))
	in
	solve_state states;;

let print_move move =
	match move with
		| Start -> "Start"
		| Right -> "right"
		| Left -> "left"
		| Up_Right -> "up-right"
		| Up_Left -> "up-left"
		| Down_Right -> "down-right"
		| Down_Left -> "down-left";;

let print_idx idx =
	let (col, row) = to_col_row idx in
	sprintf "(%d, %d)" row col;;

let rec print_history history =
		match history with
			| [] -> ()
			| record :: history' -> let () = printf "%s %s\n" (print_idx record.idx) (print_move record.move) in
		print_history history';;

let () =
	let solution = solve (int_of_string Sys.argv.(1)) in
	Deferred.upon solution
		(fun s -> print_history (List.rev s.history);(shutdown 0));
	ignore(Scheduler.go());